examples/dynamic-forall.cpp

```diff
@@ -0,0 +1,145 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-22, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include <cstdlib>
+#include <cstring>
+#include <iostream>
+
+#include "memoryManager.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+/*
+ *  Vector Addition Example with dynamic policy selection
+ *
+ *  Computes c = a + b, where a, b, c are vectors of ints.
+ *  a policy selected at run-time
+ *
+ * If CUDA is enabled, CUDA unified memory is used.
+ */
+
+//
+// Functions for checking and printing results
+//
+void checkResult(int* res, int len);
+void printResult(int* res, int len);
+
+using policy_list = camp::list<RAJA::loop_exec
+                               ,RAJA::simd_exec
+#if defined(RAJA_ENABLE_OPENMP)
+                               ,RAJA::omp_parallel_for_exec
+#endif
+#if defined(RAJA_ENABLE_CUDA)
+                               ,RAJA::cuda_exec<256>
+                               ,RAJA::cuda_exec<512>
+#endif
+                               >;
```

Comment:
Is there a way to "iterate" over the available polices and get a string id for each? Like if there was a static string in each type that we could list? Then we could make a map from string -> id that we pass to dynamic_forall, which seems nicer than just the id/maintaining the map ourselves

Reply:
Do you mean have a RAJA utility function that generates the string id?

Reply:
Perhaps have it return a list of string version of the policies? {"RAJA::loop_exec", . . ., "RAJA::cuda_exec<512>"}

Reply:
I think if we add a name property to RAJA policies that could definitely help, something like:
```
template<int N>                                                                                                                                                                                                             
struct my_cuda_exec                                                                                                                                                                                                         
{                                                                                                                                                                                                                           
  static constexpr char const *name = "my_cuda_exec";                                                                                                                                                                       
};     
``` 
However; I'm not sure how to add the number of threads to the name. Then we could compare pass in the string and make sure it matches up with the policy name.

Reply:
> Do you mean have a RAJA utility function that generates the string id?
> Perhaps have it return a list of string version list of the policies? {"RAJA::loop_exec", . . ., "RAJA::cuda_exec<512>"}

Just something that, given the camp policy list, generates a compile-time list of strings so we can say like `dynamic_forall(policy_id("cuda_exec")...`

> I think if we add a name property to RAJA policies that could definitely help, something like:

Yes!! that would help. This is something I can try to prototype in our app and get back if it's something that makes sense in raja core. thanks!

Reply:
@artv3 just googling around you might be able to hack something together for the above with a combination of https://stackoverflow.com/a/72197782 and https://stackoverflow.com/a/26824971  or use a macro for the int to string portion of things.

I also like @tomstitt suggestion of having a string we can pass in as that makes it explicit what policy a user actually wants to run.

Reply:
Thanks! We can follow up with a second pass based on user usage and then get it out of experimental.

Reply:
@rcarson3, we could do something rather simpler than that, but it still just gets us the policy name and maybe parameters, and it would be impossible to be sure we actually match what the user sees in code. Aliases, expressions etc. would be gone.  If it needs to be auto-generated, we'd probably need something like a `to_string()` for all the policies that provides an appropriate string to match the name and parameters, and then a way to compare the policies with those strings.  That's certainly doable, but it's a non-negligible amount of work, and every policy (including user supplied ones) would have to offer it.  This is why the lack of reflection in C++ is tremendously frustrating.

Another option would be to allow the user to specify a string for each policy, say instead of just taking the policy take `NamedPolicy<"name", pol>` so users can set their own arbitrary names for each policy, that would be a great deal less to support in raja and allow for custom names.  In a similar vein, we could have a way to take the offset and from that print the "pretty name" from the typeid of the policy, which would give some of the debuggability but wouldn't allow for custom strings or using the string to select a policy.

Reply:
@trws I was actually thinking of something along the lines of your `NamedPolicy<"name", pol>` in my head while I was writing up a reply, but didn't have a good idea for what that might look like.

As a user I'm okay if this sort of name addition takes a while. While it would be nice addition, it's not necessarily a must have to start prototyping with this in codes.

Reply:
If you're good with something like the wrapper, it should be a pretty simple extension that we could do soon.  Making it automatic and maintaining it for all the policies is where the significant majority of the complexity would come from.  I think either of these would be relatively straightforward:

I like this one but it requires C++20, so you might not want to use it:
```c++
using policy_list = camp::list<RAJA::NamedPolicy<"loop",RAJA::loop_exec>
                               ,RAJA::NamedPolicy<"simd",RAJA::simd_exec>>;
```

This works with c++11 and up, 17 if you want it inside a function:
```c++
static const char loop[] = "loop";
static const char simd[] = "simd";
using policy_list = camp::list<RAJA::NamedPolicy<loop,RAJA::loop_exec>
                               ,RAJA::NamedPolicy<simd,RAJA::simd_exec>>;
```

Alternately, if we take the policy by value we could construct each with a name, but there are other potential pitfalls with that.  Or we could allow a mapping from policy to string to be specified as a functor, then it could be specified as a lambda with `if constexpr` checks in it or with a little C++17 magic like this:

```c++
template<typename... Ts>
struct Overloader : Ts... {    using Ts::operator()...; };
 template<typename... T>
Overloader(T...) -> Overloader<T...>;
dynamic_forall<NamedPolicies<policies>(..., WithMapper(Overloader(
  [](loop_exec){return "loop";},
  [](simd_exec){return "simd";})));
```

Lots of options, none of them super nice sadly.  Easiest would probably if we do a version of NamedPolicy that takes classes that contain a string at `value` or something, but that's more work for users.  Pretty trivial in C++20 though, unless you're tied to icc or XL, might be alright.

Reply:
great ideas, I liked the ability to do this: `RAJA::NamedPolicy<"loop",RAJA::loop_exec>`